Applications/Piano/KeysWidget.cpp

```diff
@@ -0,0 +1,273 @@
+    "W",
+    "E",
+    "T",
+    "Y",
+    "U",
+    "O",
+    "P",
+    "]",
+};
+static const int black_key_labels_count = sizeof(black_key_labels) / sizeof(String);
+
+static const int black_key_offsets[] = {
+    white_key_width,
+    white_key_width * 2,
+    white_key_width,
+    white_key_width,
+    white_key_width * 2,
+};
+
+static const int white_key_note_accumulator[] = {
+    2,
+    2,
+    1,
+    2,
+    2,
+    2,
+    1,
+};
+
+static const int black_key_note_accumulator[] = {
+    2,
+    3,
+    2,
+    2,
+    3,
+};
+
+void KeysWidget::paint_event(GPaintEvent& event)
+{
+    GPainter painter(*this);
+    painter.translate(frame_thickness(), frame_thickness());
+
+    int note = 0;
+    int x = 0;
+    int i = 0;
+    for (;;) {
+        Rect rect(x, 0, white_key_width, frame_inner_rect().height());
+        painter.fill_rect(rect, m_key_on[note] ? note_pressed_color : Color::White);
+        painter.draw_rect(rect, Color::Black);
+        if (i < white_key_labels_count) {
+            rect.set_height(rect.height() * 1.5);
+            painter.draw_text(rect, white_key_labels[i], TextAlignment::Center, Color::Black);
+        }
+
+        note += white_key_note_accumulator[i % white_keys_per_octave];
+        x += white_key_width;
+        ++i;
+
+        if (note + m_audio_engine.octave_base() >= note_count)
+            break;
+        if (x >= frame_inner_rect().width())
+            break;
+    }
+
+    note = 1;
+    x = white_key_width - black_key_x_offset;
+    i = 0;
+    for (;;) {
+        Rect rect(x, 0, black_key_width, black_key_height);
+        painter.fill_rect(rect, m_key_on[note] ? note_pressed_color : Color::Black);
+        painter.draw_rect(rect, Color::Black);
+        if (i < black_key_labels_count) {
+            rect.set_height(rect.height() * 1.5);
+            painter.draw_text(rect, black_key_labels[i], TextAlignment::Center, Color::White);
+        }
+
+        note += black_key_note_accumulator[i % black_keys_per_octave];
+        x += black_key_offsets[i % black_keys_per_octave];
+        ++i;
+
+        if (note + m_audio_engine.octave_base() >= note_count)
+            break;
+        if (x >= frame_inner_rect().width())
+            break;
+    }
+
+    GFrame::paint_event(event);
+}
+
+static const int notes_per_white_key[] = {
+    1,
+    3,
+    5,
+    6,
+    8,
+    10,
+    12,
+};
+
+// Keep in mind that in any of these functions a note value can be out of
+// bounds. Bounds checking is done in set_key().
+
+static inline int note_from_white_keys(int white_keys)
+{
+    int octaves = white_keys / white_keys_per_octave;
+    int remainder = white_keys % white_keys_per_octave;
+    int notes_from_octaves = octaves * notes_per_octave;
+    int notes_from_remainder = notes_per_white_key[remainder];
+    int note = (notes_from_octaves + notes_from_remainder) - 1;
+    return note;
+}
+
+int KeysWidget::find_note(int x, int y) const
+{
+    if (!frame_inner_rect().contains(x, y))
```

Comment:
`x` and `y` are actually parts of LibDraw `Point`, aren't they? So why not pass it as such?

Reply:
Also a more idiomatic name for this method would be `note_for_event_position()` I believe.